model/execution/DeliveryExecutionList.php

```diff
@@ -0,0 +1,336 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\taoProctoring\model\execution;
+
+use common_ext_ExtensionsManager;
+use oat\generis\model\OntologyAwareTrait;
+use oat\oatbox\service\ConfigurableService;
+use oat\oatbox\user\User;
+use oat\tao\model\service\ApplicationService;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use oat\taoProctoring\model\monitorCache\DeliveryMonitoringService;
+use oat\taoProctoring\model\TestSessionConnectivityStatusService;
+use oat\taoQtiTest\models\SessionStateService;
+
+/**
+ * Class DeliveryHelperService
+ * @author Bartlomiej Marszal
+ */
+class DeliveryExecutionList extends ConfigurableService
+{
+    use OntologyAwareTrait;
+
+    /**
+     * Adjusts a list of delivery executions: add information, format the result
+     * @param DeliveryExecution[] $deliveryExecutions
+     * @return array
+     * @throws \common_ext_ExtensionException
+     * @throws \common_exception_Error
+     * @internal param array $options
+     */
+    public function adjustDeliveryExecutions($deliveryExecutions)
+    {
+        $executions = [];
+        $userExtraFields = $this->getUserExtraFields();
+
+        /** @var array $cachedData */
+        foreach ($deliveryExecutions as $cachedData) {
+            $executions[] = $this->getSingleExecution($cachedData, $userExtraFields);
+        }
+
+        return $executions;
+    }
+
+    /**
+     * @param $cachedData
+     * @param $userExtraFields
+     * @return array
+     * @throws \common_exception_Error
+     * @throws \common_ext_ExtensionException
+     */
+    private function getSingleExecution($cachedData, $userExtraFields)
```

Comment:
This function is still a big thing but nice with only one complexity level.
If you want to go further for something way clearer, I'd split it in parts:
- line 70 to 75: create the state
- move line 78 before the `foreach` in line 89 where it belongs to
- lines 77 to 87: create the test taker
- lines 88 to 92: create the extra fields
- lines 93 to 129: create the execution

Reply:
> lines 88 to 92: create the extra fields

Should I extract foreach to separate method?